contracts/swappers/CurveSwapper.sol

```diff
@@ -0,0 +1,257 @@
+  //                Keep names and ordering!
+  //                 Add only in the bottom.
+  // *************************************************************
+
+
+  // *************************************************************
+  //                        EVENTS
+  // *************************************************************
+
+  event Swap(
+    address pool,
+    address tokenIn,
+    address tokenOut,
+    address recipient,
+    uint priceImpactTolerance,
+    uint amountIn,
+    uint amountOut
+  );
+  // *************************************************************
+  //                        INIT
+  // *************************************************************
+
+  /// @dev Proxy initialization. Call it after contract deploy.
+  function init(address controller_) external initializer {
+    __Controllable_init(controller_);
+  }
+
+  // *************************************************************
+  //                     GOV ACTIONS
+  // *************************************************************
+
+  // *************************************************************
+  //                        PRICE
+  // *************************************************************
+
+  function getPrice(
+    address pool,
+    address tokenIn,
+    address tokenOut,
+    uint amount
+  ) public view override returns (uint) {
+    address curveMinter = _getMinter(pool);
+    (uint256 tokenInIndex, uint256 tokenOutIndex) = getTokensIndex(curveMinter, tokenIn, tokenOut);
+
+    return _callGetDY(curveMinter, tokenInIndex, tokenOutIndex, amount);
+  }
+
+  // *************************************************************
+  //                        SWAP
+  // *************************************************************
+
+  /// @dev Swap given tokenIn for tokenOut. Assume that tokenIn already sent to this contract.
+  /// @param pool Curve pool
+  /// @param tokenIn Token for sell
+  /// @param tokenOut Token for buy
+  /// @param recipient Recipient for tokenOut
+  /// @param priceImpactTolerance Price impact tolerance. Must include fees at least.
+  function swap(
+    address pool,
+    address tokenIn,
+    address tokenOut,
+    address recipient,
+    uint priceImpactTolerance
+  ) external override {
+    address curveMinter = _getMinter(pool);
+    uint amountIn = IERC20(tokenIn).balanceOf(address(this));
+
+    (uint256 tokenInIndex, uint256 tokenOutIndex) = getTokensIndex(curveMinter, tokenIn, tokenOut);
+
+    // scope for checking price impact
+    uint256 priceBefore = getPrice(pool, tokenIn, tokenOut, amountIn);
+
+    require(IERC20(tokenIn).balanceOf(address(this)) >= amountIn, "Wrong amountIn");
+    IERC20(tokenIn).approve(curveMinter, amountIn);
+
+    uint256 amountOut = _callExchange(curveMinter, tokenInIndex, tokenOutIndex, amountIn, _LIMIT);
+    require(IERC20(tokenOut).balanceOf(address(this)) >= amountOut, "Wrong amountOut");
+
+    uint256 priceAfter = getPrice(pool, tokenIn, tokenOut, amountIn);
```

Comment:
1) getPrice pretty expensive, no need to call it twice
2) you need to check what you exactly received, calculate the price from received amounts

Reply:
fixed